### PR TITLE
Allow creation of RefCountable objects via Make()

### PR DESCRIPTION
**Pull request by yadij**
Compared to calling new() directly, using a Pointer-returning Make()
reduces memory leaks related to freshly allocated heap objects. Perfect
forwarding of constructor parameters through Make() minimizes overhead.

    const auto foo1 = new Foo(...); // XXX: leak-prone
    const auto foo2 = Foo::Pointer::Make(...); // OK

Future changes will prohibit RefCountable object creation via direct
new() calls and require using Make() or an equivalent safety API.

### CODE REVIEW

**Comment by yadij** [comment]
TODO list:
 * convert RefCountable classes from `new Foo` to `FooPointer::Make()`
 * locate code depending on implicit conversion via raw-pointer constructor
 * make raw-pointer constructor private

File: src/pconn.cc
Comment (yadij) [reply src/pconn.cc:174]: Will do if they build clean after hiding the raw-pointer constructor. For now there are too many warning/errors that come up to see clearly.

File: src/CommCalls.cc
Comment (yadij) [reply src/CommCalls.cc:36]: Okay. I can do them in a separate PR if you insist, but these will need doing either way before this PR can go in.

FTR; They are needed along with some `RefCount` operators specializing the use of `nullptr_t` parameters before the raw-pointer constructor can be properly hidden in a private section.

File: src/base/RefCount.h
Comment (yadij) [reply src/base/RefCount.h:32]: Aha, sweet. That is one new C++ feature I hadn't looked into uses for yet. Seems to work perfectly (pun intended).

**Comment by yadij** [comment]
> please do not convert any more classes until we develop consensus regarding this PR.

Ok.

FWIW, the classes done so far are test/example of the various cases the `RefCount` API has to deal with.
 * `MemBlob` is typical simple class.
 * `Comm::Connection` exercises a mix of construct and conversion operators.
 * `MasterXaction` has custom `Make*()` construction

In my test builds I also see `HttpReplyPointer` which exercises a bunch of cases where the `RefCount` is inherited from a parent class, and has initializer-list construction of `RefCount` members. I was intending to tackle those today and then get you to do a deep review before delving into the remainder.

File: src/MasterXaction.h
Comment (rousskov) [reply src/MasterXaction.h:77]: > This `friend` was needed so the `RefCount::Make()` template can have access to the hidden `MasterXAction` constructor (when used by the `MasterXaction::Make*()` methods). 

Yes, I know. To avoid misunderstanding, the point of this change request is that existing MasterXaction::Make*() methods should not use the new RefCount::Make(). Instead, the existing two custom MasterXaction makers should be preserved without any changes (including changes that create compiler errors that, in turn, require adding this friend declaration).

File: src/pconn.cc
Comment (rousskov) [reply src/pconn.cc:174]: Sounds good. FWIW, if this code does not "build clean" without any changes, before or after hiding, then I would suspect that something else is broken. Creating a nil smart pointer should be allowed (and happen by default).

**Review by rousskov — COMMENTED** [review]
> In my test builds I also see HttpReplyPointer which exercises a bunch of cases where the RefCount is inherited from a parent class, and has initializer-list construction of RefCount members. I was intending to tackle those today and then get you to do a deep review before delving into the remainder.

I do _not_ recommend throwing in HttpReplyPointer, at least until this PR is effectively ready to be merged. We have enough complexity to deal with already IMHO, and that class might introduce a whole new set of problems that deserve dedicated attention.

File: src/base/RefCount.h
Comment (rousskov) [reply src/base/RefCount.h]: Not done AFAICT: This PR code is still copying (rather than perfectly forwarding) because it is still not using references. If the earlier wrapper() example was not helpful, perhaps this text about forwarding references will be: https://en.cppreference.com/w/cpp/language/reference#Forwarding_references

Please let me know if I should provide a specific code replacement instead. I just thought that showing where to insert a few magic symbols would jeopardize understanding of the underlying reasoning/mechanism (which is useful in other contexts as well).

Comment (yadij) [reply src/base/RefCount.h]: I have rebased since, it is not in the latest [24727e6](https://github.com/squid-cache/squid/pull/1458/commits/24727e6ad0d58d6c4788c6c14e8abc4612846744)

Comment (yadij) [reply src/base/RefCount.h]: Hmm, if I am understanding correctly you are suggesting to force the use of rvalue references instead of preserving the caller references (or lack of) ?
```c++
    inline static auto Make(Args&&... args) {
        return RefCount<C>(new C(std::forward<Args>(args)...));
    }
```

Comment (rousskov) [reply src/base/RefCount.h]: > if I am understanding correctly you are suggesting to force the use of rvalue references instead of preserving the caller references (or lack of)?

I would not describe my suggestion as forcing the use of rvalue references. And following my suggestion would also preserve caller references, so that desirable preservation cannot be accurately described as "instead".

Perfect forwarding requires [forwarding references](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references), which are not equivalent to rvalue references: Forwarding references use rvalue reference _declaration syntax_, but forwarding references effect/semantics is different from rvalue references. Not all references that use rvalue reference declaration syntax are forwarding references (because forwarding references require special/extra preconditions). For example, a classic move constructor uses an rvalue reference but not a forwarding reference. Due to those semantic differences, I believe that calling a forwarding reference an rvalue reference is also technically inaccurate, but since they share the same rvalue reference declaration syntax, I suspect that does happen a lot.

The updated branch code uses forwarding references, so I am resolving this change request, but if you would like to discuss any related concerns, I am happy to participate. Personally, I find this C++ concept difficult to master, and Herb Sutter [calls](https://stackoverflow.com/questions/29590883/herb-sutters-cppcon-perfect-forwarding-slides) it "unteachable", but since it is so useful, I believe that mastering it is worth the effort.

P.S. Forwarding references were previously known as "universal references". Some of the primary discussions about them can be found using that earlier term.

**Review by rousskov — CHANGES_REQUESTED** [review]
Great progress, thank you!

File: src/base/RefCount.h
Comment (yadij) [reply src/base/RefCount.h:106]: Er, the original constructor was `C const *` . Highlighting why this style of `const` positioning is the safer one .. .we can see more easily when it is wrong.

File: src/base/RefCount.h
Comment (yadij) [reply src/base/RefCount.h:32]: IIRC we discussed this in the meeting last week. Prohibiting conversion from raw-pointer to Pointer for the `RefCountable` classes avoids all that extra annoyance with `new`.

FWIW, I intend to to make this constructor public again with a TODO for the transition once this PR passes review.

Comment (rousskov) [reply src/base/RefCount.h:32]: > IIRC we discussed this in the meeting last week. Prohibiting conversion from raw-pointer to Pointer for the `RefCountable` classes avoids all that extra annoyance with `new`.

We did discuss the dangers of leaving `new` result in a raw pointer. I believe we agree that those dangers should be avoided. I hope we also agree that Make() is the right solution to avoid those dangers. However, all that does not imply that we should hide this conversion constructor.

```C++
C *rawP = new C; // a problem (should be prohibited but would be allowed, in some contexts, by this change)
Pointer p1 = new C; // not really a problem (but would be prohibited by this change)
Pointer p2 = rawP; // not really a problem (but would be prohibited by this change) or a minor out-of-scope problem
```
 
> FWIW, I intend to to make this constructor public again with a TODO for the transition once this PR passes review.

Noted, but we do not want to transition to a hidden conversion constructor, at least not as far as this PR, this problem, and Make() solutions are concerned. All those things require a very different transition goal and steps, the ones outlined in this change request.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/base/RefCount.h
Comment (yadij) [reply src/base/RefCount.h:32]: Sigh. I thought we had this settled.

The problem I am solving is that mixing explicit `rawP = new C;  ... ; Pointer p1 = rawP; ...` has problems for us determining whether `rawP` needs developers to code explicit `delete`(s) [in extra code paths handling `rawP` copies] or if the `Pointer` takes full control and does it implicitly in all code.

By explicitly prohibiting conversion between the types of pointer semantic, we clarify for reviewers and developers which memory management paradigm to apply to the variables (C++ RAII vs. C-style). Making the review process easier and requirements simpler to inform new developers about.


Essentially removing `new'/'delete' from the C++ language is an entirely different long-term situation which I am not going near tackling in this PR or any of its direct followups. I am leaving that to you after this transition converts most `RefCountable` child classes to `Make()` and highlights the ones with issues.

Comment (rousskov) [reply src/base/RefCount.h:32]: > The problem I am solving is that mixing explicit `rawP = new C; ... ; Pointer p1 = rawP; ...` has problems for us determining whether `rawP` needs developers to code explicit `delete`(s) [in extra code paths handling `rawP` copies] or if the `Pointer` takes full control and does it implicitly in all code.

The `rawP = new C; ... ; Pointer p1 = rawP` problem is solved by migrating to Make(). After that migration, `rawP = new C` disappears completely. Hiding this conversion constructor does not help with that migration and does not enforce that migration, but creates other problems. Some of those other problems are rather complex/controversial. _This_ PR does not need to open that Pandora box.

Migration from `const auto rawP = new C()` to the equivalent of `const auto p = Make()` does not solve all pointer-related problems, of course. If you would like to solve some other specific problem, let's discuss, but that discussion does not need to block this PR AFAICT _and_ should be based on Make()-using code (as if this PR has been merged).


> By explicitly prohibiting conversion between the types of pointer semantic, we clarify for reviewers and developers which memory management paradigm to apply to the variables (C++ RAII vs. C-style). Making the review process easier and requirements simpler to inform new developers about.

Agreed. Implicit conversions do have significant costs. On the other hand, we should not prohibit all useful (and occasionally essential) code that uses raw pointers correctly. Raw pointers are the best solution for some problems. Again, I am happy to discus solutions for reducing incorrect use of raw pointers in post-Make() code. However, I insist on formulation of those problems (if they are to be discussed) using post-Make() code templates/examples. The current example quoted above is pre-Make(). Such examples will not exist (i.e. become invalid) after we convert to Make().

AFAICT, there is basic consensus regarding conversion to Make(), so we should assume or capitalize on that great simplification/improvement when discussing further pointer-related improvements.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/base/RefCount.h
Comment (rousskov) [inline src/base/RefCount.h:30]: Just to document the new function template. We do not have to do that if you think it is obvious:

```suggestion
    /// creates a new C object using given C constructor arguments (if any)
    /// \returns a refcounting pointer to the created object
    template<typename... Args>
```

File: src/base/RefCount.h
Comment (rousskov) [inline src/base/RefCount.h:32]: `std::forward` is declared in [utility](https://en.cppreference.com/w/cpp/header/utility). We should `#include` that header explicitly instead of waiting for that inclusion to become necessary.

**Review by rousskov — DISMISSED** [review]
LGTM with minor polishing touches. I also edited the PR title/description, primarily to remove minor technical inaccuracies and document planned enforcement. Please feel free to merge without additional reviews if you agree with those adjustments.

I would prefer to see one or two classes (the ones that are easy to convert without any controversies) converted to use Make() in this PR, so that we are not merging unused code, but I do not insist on that.

**Review by rousskov — APPROVED** [review]

**Pull request closed without merging**
